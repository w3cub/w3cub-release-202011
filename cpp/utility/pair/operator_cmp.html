
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Operators (Std&#58;&#58;pair) - C++ - W3cubDocs</title>
  
  <meta name="description" content=" The &#60;, &#60;&#61;, &#62;, &#62;&#61;, and !&#61; operators are synthesized from operator&#60;&#61;&#62; and operator&#61;&#61; respectively. ">
  <meta name="keywords" content="operator, !, operators, std, pair, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/pair/operator_cmp.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">operator==,!=,&lt;,&lt;=,&gt;,&gt;=,&lt;=&gt;<small>(std::pair)</small>
</h1>            <table class="t-dcl-begin"> <tr class="t-dsc-header"> <th> Defined in header <code><a href="../../header/utility" title="cpp/header/utility">&lt;utility&gt;</a></code> </th> <th> </th> <th> </th> </tr>  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(1)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator==( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator==( const std::pair&lt;T1, T2&gt;&amp; lhs,
                           const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(2)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator!=( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator!=( const std::pair&lt;T1, T2&gt;&amp; lhs,
                           const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(3)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator&lt;( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator&lt;( const std::pair&lt;T1, T2&gt;&amp; lhs,
                          const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(4)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator&lt;=( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator&lt;=( const std::pair&lt;T1, T2&gt;&amp; lhs,
                           const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(5)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator&gt;( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator&gt;( const std::pair&lt;T1, T2&gt;&amp; lhs,
                          const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(6)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx14"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
bool operator&gt;=( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx14">(until C++14)</span> </td> </tr> <tr class="t-dcl t-since-cxx14 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr bool operator&gt;=( const std::pair&lt;T1, T2&gt;&amp; lhs,
                           const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx14">(since C++14)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>  <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template&lt; class T1, class T2, class U1, class U2 &gt;
constexpr std::common_comparison_category_t&lt;synth-three-way-result&lt;T1, U1&gt;,
                                            synth-three-way-result&lt;T2, U2&gt;&gt;
    operator&lt;=&gt;( const std::pair&lt;T1, T2&gt;&amp; lhs, const std::pair&lt;U1, U2&gt;&amp; rhs );</pre>
</td> <td> (7) </td> <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>  </table> <div class="t-li1">
<span class="t-li">1-2)</span> Tests if both elements of <code>lhs</code> and <code>rhs</code> are equal, that is, compares <code>lhs.first</code> with <code>rhs.first</code> and <code>lhs.second</code> with <code>rhs.second</code>.</div> <pre data-language="cpp">operator&lt;</pre> <div class="t-li1">
<span class="t-li">7)</span> Compares <code>lhs</code> and <code>rhs</code> lexicographically by <a href="../../standard_library/synth-three-way" title="cpp/standard library/synth-three-way"><code>synth-three-way</code></a>, that is, compares the first elements and only if they are equivalent, compares the second elements. <a href="../../standard_library/synth-three-way" title="cpp/standard library/synth-three-way"><code>synth-three-way-result</code></a> is the return type of <code><i>synth-three-way</i></code>.</div> <table class="t-rev-begin"> <tr class="t-rev t-since-cxx20">
<td> <p>The <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code>, <code>&gt;=</code>, and <code>!=</code> operators are <a href="../../language/operators#Comparison_operators" title="cpp/language/operators">synthesized</a> from operator<span class="sy1">&lt;=&gt;</span> and operator<span class="sy1">==</span> respectively.</p>
</td> <td><span class="t-mark-rev t-since-cxx20">(since C++20)</span></td>
</tr> </table>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs, rhs </td> <td> - </td> <td> pairs to compare </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> <code>true</code> if both <code>lhs.first == rhs.first</code> and <code>lhs.second == rhs.second</code>, otherwise <code>false</code>
</div> <div class="t-li1">
<span class="t-li">2)</span> <code>!(lhs == rhs)</code>
</div> <div class="t-li1">
<span class="t-li">3)</span> If <code>lhs.first &lt; rhs.first</code>, returns <code>true</code>. Otherwise, if <code>rhs.first &lt; lhs.first</code>, returns <code>false</code>. Otherwise, if <code>lhs.second &lt; rhs.second</code>, returns <code>true</code>. Otherwise, returns <code>false</code>.</div> <div class="t-li1">
<span class="t-li">4)</span> <code>!(rhs &lt; lhs)</code>
</div> <div class="t-li1">
<span class="t-li">5)</span> <code>rhs &lt; lhs</code>
</div> <div class="t-li1">
<span class="t-li">6)</span> <code>!(lhs &lt; rhs)</code>
</div> <div class="t-li1">
<span class="t-li">7)</span> <code><a href="../../standard_library/synth-three-way" title="cpp/standard library/synth-three-way">synth-three-way</a>(lhs.first, rhs.first)</code> if it is not equal to <code>​0​</code>, otherwise <code><a href="../../standard_library/synth-three-way" title="cpp/standard library/synth-three-way">synth-three-way</a>(lhs.second, rhs.second)</code>.</div> <h3 id="Example"> Example</h3> <div class="t-example">
<p>Because operator<span class="sy1">&lt;</span> is defined for pairs, containers of pairs can be sorted.</p>
<div class="cpp source-cpp"><pre data-language="cpp">#include &lt;algorithm&gt;
#include &lt;iomanip&gt;
#include &lt;iostream&gt;
#include &lt;string&gt;
#include &lt;utility&gt;
#include &lt;vector&gt;
 
int main()
{
    std::vector&lt;std::pair&lt;int, std::string&gt;&gt; v = {{2, "baz"},
                                                  {2, "bar"},
                                                  {1, "foo"}};
    std::sort(v.begin(), v.end());
 
    for (auto p: v)
        std::cout &lt;&lt; "{" &lt;&lt; p.first &lt;&lt; ", " &lt;&lt; std::quoted(p.second) &lt;&lt; "}\n";
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">{1, "foo"}
{2, "bar"}
{2, "baz"}</pre></div> </div> <h3 id="Defect_reports"> Defect reports</h3> <p>The following behavior-changing defect reports were applied retroactively to previously published C++ standards.</p>
<table class="dsctable"> <tr> <th> DR </th> <th> Applied to </th> <th> Behavior as published </th> <th> Correct behavior </th>
</tr> <tr> <td> <a rel="nofollow" class="external text" href="https://cplusplus.github.io/LWG/issue296">LWG 296</a> </td> <td> C++98 </td> <td> the descriptions of operators other than <code>==</code> and <code>&lt;</code> were missing </td> <td> added </td>
</tr> <tr> <td> <a rel="nofollow" class="external text" href="https://cplusplus.github.io/LWG/issue3865">LWG 3865</a> </td> <td> C++98 </td> <td> comparison operators only accepted <code>pair</code>s of the same type </td> <td> accept <code>pair</code>s of different types </td>
</tr>
</table> <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="../tuple/operator_cmp" title="cpp/utility/tuple/operator cmp"> <span class="t-lines"><span>operator==</span><span>operator!=</span><span>operator&lt;</span><span>operator&lt;=</span><span>operator&gt;</span><span>operator&gt;=</span><span>operator&lt;=&gt;</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-since-cxx20">(C++20)</span></span></span></div> </td> <td> lexicographically compares the values in the tuple <br> <span class="t-mark">(function template)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    © cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/pair/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/pair/operator_cmp</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
