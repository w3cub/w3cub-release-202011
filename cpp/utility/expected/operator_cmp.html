
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Operator&#61;&#61;(std&#58;&#58;expected) - C++ - W3cubDocs</title>
  
  <meta name="description" content=" Performs comparison operations on expected objects. ">
  <meta name="keywords" content="operator, std, expected, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/expected/operator_cmp.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">operator==<small>(std::expected)</small>
</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2, class E2 &gt;
  requires (!std::is_void_v&lt;T2&gt;)
friend constexpr bool operator==( const expected&amp; lhs,
                                  const std::expected&lt;T2, E2&gt;&amp; rhs );</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is not <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2, class E2 &gt;
  requires std::is_void_v&lt;T2&gt;
friend constexpr bool operator==( const expected&amp; lhs,
                                  const std::expected&lt;T2, E2&gt;&amp; rhs );</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2 &gt;
friend constexpr bool operator==( const expected&amp; x, const T2&amp; val );</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is not <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class E2 &gt;
friend constexpr bool operator==( const expected&amp; x,
                                  const unexpected&lt;E2&gt;&amp; e );</pre>
</td> <td> (4) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>Performs comparison operations on <code>expected</code> objects.</p>
<span class="t-li">1,2)</span> Compares two <code>expected</code> objects. The objects compare equal if and only if <code>lhs.has_value()</code> and <code>rhs.has_value()</code> are equal, and the contained values are equal. <ul>
<li> For overload <span class="t-v">(1)</span>, if the expressions <code>*lhs == *rhs</code> and <code>lhs.error() == rhs.error()</code> are not well-formed, or if their results are not convertible to <code>bool</code>, the program is ill-formed. </li>
<li> For overload <span class="t-v">(2)</span>, if the expressions <code>lhs.error() == rhs.error()</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li>
</ul> <span class="t-li">3)</span> Compares <code>expected</code> object with a value. The objects compare equal if and only if <code>x</code> contains an expected value, and the contained value is equal to <code>val</code>. <ul><li> If the expression <code>*x == val</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li></ul> <span class="t-li">4)</span> Compares <code>expected</code> object with an unexpected value. The objects compare equal if and only if <code>x</code> contains an unexpected value, and the contained value is equal to <code>e.error()</code>. <ul><li> If the expression <code>x.error() == e.error()</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li></ul> <p>These functions are not visible to ordinary <a href="../../language/unqualified_lookup" title="cpp/language/unqualified lookup">unqualified</a> or <a href="../../language/qualified_lookup" title="cpp/language/qualified lookup">qualified lookup</a>, and can only be found by <a href="../../language/adl" title="cpp/language/adl">argument-dependent lookup</a> when <code>std::expected&lt;T, E&gt;</code> is an associated class of the arguments.</p>
<p>The <code>!=</code> operator is <a href="../../language/operators#Relational_operators" title="cpp/language/operators">synthesized</a> from <code>operator==</code>.</p>
<h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs, rhs, x </td> <td> - </td> <td> <code>expected</code> object to compare </td>
</tr> <tr class="t-par"> <td> val </td> <td> - </td> <td> value to compare to the expected value contained in <code>x</code> </td>
</tr> <tr class="t-par"> <td> e </td> <td> - </td> <td> value to compare to the unexpected value contained in <code>x</code> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> If <code>lhs.has_value() != rhs.has_value()</code>, returns <code>false</code>. Otherwise, if <code>lhs.has_value()</code> is true, returns <code>*lhs == *rhs</code>. Otherwise, returns <code>lhs.error() == rhs.error()</code>.</div> <div class="t-li1">
<span class="t-li">2)</span> If <code>lhs.has_value() != rhs.has_value()</code>, returns <code>false</code>. Otherwise, if <code>lhs.has_value()</code> is true, returns <code>true</code>. Otherwise, returns <code>lhs.error() == rhs.error()</code>.</div> <div class="t-li1">
<span class="t-li">3)</span> Returns <code>x.has_value() &amp;&amp; static_cast&lt;bool&gt;(*x == val)</code>.</div> <div class="t-li1">
<span class="t-li">4)</span> Returns <code>!x.has_value() &amp;&amp; static_cast&lt;bool&gt;(x.error() == e.error())</code>.</div> <h3 id="Exceptions"> Exceptions</h3> <p>Throws when and what the comparison throws.</p>
<h3 id="Example"> Example</h3>  <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="unexpected" title="cpp/utility/expected/unexpected"> <span class="t-lines"><span>unexpected</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> represented as an unexpected value <br> <span class="t-mark">(class template)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    Â© cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/expected/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/expected/operator_cmp</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
